tidy3d/plugins/near2far/near2far.py

```diff
@@ -1,115 +1,373 @@
+
+            idx_uv = [0, 2]
+            tangent_fields = ['x', 'z']
+            normal_field = 'y'
+            signs = [1.0, -1.0]
+
+        else:
+
+            idx_uv = [0, 1]
+            tangent_fields = ['x', 'y']
+            normal_field = 'z'
+            signs = [-1.0, 1.0]
+
+        if self.normal_dir == '-':
+            signs = [-1.0 * i for i in signs]
+
+        # compute surface current densities and delete unneeded field components
+        currents = field_data.copy(deep=True)
+
+        currents.data_dict['J'+tangent_fields[1]] = currents.data_dict.pop('H'+tangent_fields[0])
+        currents.data_dict['J'+tangent_fields[0]] = currents.data_dict.pop('H'+tangent_fields[1])
+        del currents.data_dict['H'+normal_field]
+
+        currents.data_dict['M'+tangent_fields[1]] = currents.data_dict.pop('E'+tangent_fields[0])
+        currents.data_dict['M'+tangent_fields[0]] = currents.data_dict.pop('E'+tangent_fields[1])
+        del currents.data_dict['E'+normal_field]
 
+        currents.data_dict['J'+tangent_fields[0]].values *= signs[0]
+        currents.data_dict['J'+tangent_fields[1]].values *= signs[1]
+
+        currents.data_dict['M'+tangent_fields[0]].values *= signs[1]
+        currents.data_dict['M'+tangent_fields[1]].values *= signs[0]
+
+        # colocate surface currents on a regular grid of points on the monitor based on wavelength
+        colocation_points = [None] * 3
+        colocation_points[self.axis] = self.mon.center[self.axis]
+
+        for idx in idx_uv:
+
+            # pick sample points on the monitor and handle the possibility of an "infinite" monitor
+            start = np.maximum(
+                self.mon.center[idx] - self.mon.size[idx] / 2.0,
+                self.mon.center[idx] - self.sim_data.simulation.size[idx] / 2.0
+                )
+            stop = np.minimum(
+                self.mon.center[idx] + self.mon.size[idx] / 2.0,
+                self.mon.center[idx] + self.sim_data.simulation.size[idx] / 2.0
+                )
+            size = stop - start
+
+            num_pts = int(np.ceil(pts_per_wavelength * size / wavelength))
+            points = np.linspace(start, stop, num_pts)
+            colocation_points[idx] = points
+
+        currents = currents.colocate(*colocation_points)
         try:
-            Ex = field_data.Ex.sel(f=frequency)
-            Ey = field_data.Ey.sel(f=frequency)
-            # self.Ez = field_data['Ez'].sel(f=frequency)
-            Hx = field_data.Hx.sel(f=frequency)
-            Hy = field_data.Hy.sel(f=frequency)
+            self.currents = currents.sel(f=frequency)
         except Exception as e:
             raise SetupError(
-                f"Frequency {frequency} not found in all fields " f"from monitor '{mon_name}'."
+                f"Frequency {frequency} not found in fields for monitor '{self.mon.name}'."
             ) from e
 
-        self.k0 = 2 * np.pi * frequency / C_0
+class Near2Far(Tidy3dBaseModel):
```

Comment:
`pydantic.BaseModel`

Reply:
For this to work I added a `Config` class with `arbitrary_types_allowed = True` to handle xarray

Reply:
great. thanks.  eventually we may move these to Tidy3dBaseModel but let's just keep it generic for now.